### PR TITLE
Devices and external libraries options, make install, Makefile fix 

### PR DESCRIPTION
**Pull request by nicolaspi**
- New features that allow to generate code according to operator's devices compatibility (cpu and/or gpu), and to specify external libraries to be linked.
- The Makefile has been fixed to work with the latest tensorflow version (tested on the version 1.3 built from source).
- Added 'make install' that copies the built library to the tensorflow's library directory.

### CODE REVIEW

**Comment by nicolaspi** [comment]
Looks like a version issue with tensorflow, the linker does not find  `libtensorflow_framework.so` which was introduced by [this](https://github.com/tensorflow/tensorflow/commit/5c7f9e316d8c7735308a217310350d416d7498cc) commit (19 days ago).

**Comment by sjperkins** [comment]
Thanks for the PR. It looks like the  new `libtensorflow_framework.so` linking style is going into tensorflow 1.4.0 (see https://github.com/tensorflow/tensorflow/issues/13607 for e.g.). It would probably be best to wait for this release before merging this and releasing a new tfopgen.

File: tfopgen/create_op.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
 def make_template_kwargs(op_name, py_op_name,
                          project, library, op_inputs, op_outputs,
-                         op_type_attrs, op_other_attrs, op_doc):
+                         op_type_attrs, op_other_attrs, op_doc, op_libs, op_devices_comps, op_abi):
```
Comment (sjperkins) [inline tfopgen/create_op.py:11]: Move `op_libs, op_devices_comps, op_abi` to a newline.

File: README.rst [hunk 2]
```diff
@@ -73,6 +75,12 @@ For example, we can define the outline for a ``ComplexPhase`` operator in the ``
         (2) of (L, M) sky coordinates with shape (nsrc, 2)
         (3) of frequencies,
       compute the complex phase with shape (nsrc, ntime, nbl, nchan)
+    devices_compatibilities:
+        - "cpu"
+        - "gpu"
+    extra_libraries_deps:
+        - "m"
+    use_old_abi: yes
```
Comment (sjperkins) [inline README.rst:84]: I think if we start mixing the Makefile options with the operator options, the yaml will start to be cluttered.

How about a `makefile` section to distinguish it from the rest of the yaml?

```yaml
makefile:
    use_old_abi: yes
    libraries:
       - "m"
```

File: README.rst [hunk 3]
```diff
@@ -73,6 +75,12 @@ For example, we can define the outline for a ``ComplexPhase`` operator in the ``
         (2) of (L, M) sky coordinates with shape (nsrc, 2)
         (3) of frequencies,
       compute the complex phase with shape (nsrc, ntime, nbl, nchan)
+    devices_compatibilities:
```
Comment (sjperkins) [inline README.rst:78]: Rename `devices_compatibilities` to `devices`

File: README.rst [hunk 4]
```diff
@@ -49,6 +49,8 @@ The user should provide a YAML configuration file defining the operator:
 -  polymorphic type attributes.
 -  other attributes.
 -  documentation.
+-  devices compatibilities.
+-  optional external libraries.
```
Comment (sjperkins) [inline README.rst:53]: `device compatibilities` to `devices`
`optional external libraries` to `makefile` options

File: README.rst [hunk 5]
```diff
@@ -147,6 +155,33 @@ generator code as the range of attribute behaviour is complex.
     op_other_attrs:
         - "iterations: int32 >= 2",
 
+Sources codes will be generated according to the specified devices.
```
Comment (sjperkins) [inline README.rst:158]: `Source code will be generated for the requested devices`

File: README.rst [hunk 6]
```diff
@@ -73,6 +75,12 @@ For example, we can define the outline for a ``ComplexPhase`` operator in the ``
         (2) of (L, M) sky coordinates with shape (nsrc, 2)
         (3) of frequencies,
       compute the complex phase with shape (nsrc, ntime, nbl, nchan)
+    devices_compatibilities:
+        - "cpu"
+        - "gpu"
+    extra_libraries_deps:
```
Comment (sjperkins) [inline README.rst:81]: `extra_library_deps` to `libraries`

File: tfopgen/examples/get_slide_patch.yml [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  - ["tif_files: string", [null,1]] # (file_paths)
+  - ["levels: int32", [null, 1]] # (levels)
+outputs:
+  - ["images: int32", [null, null, null, 4]] # (N, w,h,(ARGB))
+doc: >
+  Get Slide patch
+extra_libraries_deps:
+  - "openslide"
+devices_compatibilities:
+  - "cpu"
```
Comment (sjperkins) [inline tfopgen/examples/get_slide_patch.yml:19]: Update with makefile section

File: tfopgen/templates/test_source.j2 [hunk 8]
```diff
@@ -24,7 +24,7 @@ class Test{{op_name}}(unittest.TestCase):
         {% set comma = last_joiner(",") -%}
         type_permutations = [
         {%- for perm in op_np_type_perms %}
-        {{ perm | replace("'", "") | indent(4, True) }}{{comma(loop.last)}}
+        {{ perm[0] | replace("'", "") | indent(4, True) }}{{comma(loop.last)}}
```
Comment (sjperkins) [inline tfopgen/templates/test_source.j2:27]: What does this do?

File: README.rst [hunk 9]
```diff
@@ -147,6 +155,33 @@ generator code as the range of attribute behaviour is complex.
...
+
+    extra_libraries_deps:
+        - "opencv"
+        - "jpeg"
+
+Note on ``gcc`` version ``>=5``: gcc uses the new C++ ABI since version 5.
+The binary pip packages available on the TensorFlow website are built
+with gcc4 that uses the older ABI. Set the option ``use_old_abi`` with the value ``yes``
+to allow the library to be compatible with the older abi.
+
```
Comment (sjperkins) [inline README.rst:178]: Change to:

```
Note that gcc version 5 and above uses a new C++ ABI.
However, the official tensorflow binary pip packages are built with gcc 4
which uses an older ABI. Set the option `use_old_abi: yes` to make
the library compatible with the older ABI.

File: tfopgen/templates/Makefile.j2 [hunk 10]
```diff
@@ -21,12 +22,12 @@ OBJECTS=$(addsuffix .o, $(basename $(SOURCES)))
...
 
 # Compiler flags
-INCLUDES= -I $(TF_INC)
+INCLUDES= -I $(TF_INC) -I$(TF_INC)/external/nsync/public
 CPPFLAGS=-std=c++11 $(TF_FLAGS) $(INCLUDES) -fPIC -fopenmp -O2 -march=native -mtune=native
 NVCCFLAGS=-std=c++11 -D GOOGLE_CUDA=$(TF_CUDA) $(TF_FLAGS) $(INCLUDES) \
 	-x cu --compiler-options "-fPIC" --gpu-architecture=sm_30 -lineinfo
 
-LDFLAGS = -fPIC -fopenmp
+LDFLAGS = -fPIC -fopenmp -L$(TF_LIB) -ltensorflow_framework{{extra_libs}}
```
Comment (sjperkins) [inline tfopgen/templates/Makefile.j2:31]: One possible way of handling the the presence or absence of the `-ltensorflow_framework`   library would be to add `tensorflow_version` to the YAML.

**Review by sjperkins — CHANGES_REQUESTED** [review]

**Pull request closed without merging**